In [1]:
import numpy as np

## KNN Regression Algorithm

In [3]:
data = np.genfromtxt("../Dataset/diabetes.csv", delimiter=',')
data[:5]

array([[ 3.80759064e-02,  5.06801187e-02,  6.16962065e-02,
         2.18723550e-02, -4.42234984e-02, -3.48207628e-02,
        -4.34008457e-02, -2.59226200e-03,  1.99084209e-02,
        -1.76461252e-02,  1.51000000e+02],
       [-1.88201653e-03, -4.46416365e-02, -5.14740612e-02,
        -2.63278347e-02, -8.44872411e-03, -1.91633397e-02,
         7.44115641e-02, -3.94933829e-02, -6.83297436e-02,
        -9.22040496e-02,  7.50000000e+01],
       [ 8.52989063e-02,  5.06801187e-02,  4.44512133e-02,
        -5.67061055e-03, -4.55994513e-02, -3.41944659e-02,
        -3.23559322e-02, -2.59226200e-03,  2.86377052e-03,
        -2.59303390e-02,  1.41000000e+02],
       [-8.90629394e-02, -4.46416365e-02, -1.15950145e-02,
        -3.66564468e-02,  1.21905688e-02,  2.49905934e-02,
        -3.60375700e-02,  3.43088589e-02,  2.26920226e-02,
        -9.36191133e-03,  2.06000000e+02],
       [ 5.38306037e-03, -4.46416365e-02, -3.63846922e-02,
         2.18723550e-02,  3.93485161e-03,  1.55961395e-02,
  

In [9]:
my_list = data.tolist()
print(my_list[:5])

[[0.0380759064334241, 0.0506801187398187, 0.0616962065186885, 0.0218723549949558, -0.0442234984244464, -0.0348207628376986, -0.0434008456520269, -0.00259226199818282, 0.0199084208763183, -0.0176461251598052, 151.0], [-0.00188201652779104, -0.044641636506989, -0.0514740612388061, -0.0263278347173518, -0.00844872411121698, -0.019163339748222, 0.0744115640787594, -0.0394933828740919, -0.0683297436244215, -0.09220404962683, 75.0], [0.0852989062966783, 0.0506801187398187, 0.0444512133365941, -0.00567061055493425, -0.0455994512826475, -0.0341944659141195, -0.0323559322397657, -0.00259226199818282, 0.00286377051894013, -0.0259303389894746, 141.0], [-0.0890629393522603, -0.044641636506989, -0.0115950145052127, -0.0366564467985606, 0.0121905687618, 0.0249905933641021, -0.0360375700438527, 0.0343088588777263, 0.0226920225667445, -0.0093619113301358, 206.0], [0.00538306037424807, -0.044641636506989, -0.0363846922044735, 0.0218723549949558, 0.00393485161259318, 0.0155961395104161, 0.00814208360519

### Randomize arrangement of data

In [10]:
import random
random.shuffle(my_list)

### Split into training, validation, and test

In [12]:
train = []
test = []
valid = []

for item in my_list:
    test_val = random.uniform(0, 1.0)
    if test_val <= 0.7:
        train.append(item)
    elif test_val <= 0.85:
        valid.append(item)
    else:
        test.append(item)

In [13]:
print(len(train), len(valid), len(test))

303 56 83


In [85]:
def KNNRegression(test,valid,k=5):
    error=0
    distances = []
    for validation in valid:
        for training in train:
            euclidean = sum([(p-q) ** 2 for p, q in zip(validation[:-1], training[:-1])])
            distances.append((euclidean, training))
        distances.sort(key = lambda item: item[0])
        samples = distances[:k]
        
        for sample in samples:
            data = sample[1]
            data = data[-1:]
            data.append(data[0])
        
        avg_o = sum(data)/k
        
        error += ((validation[-1:][0] - avg_o)**2)
        
        distances = []
        samples = []
    
    return error / len(valid)

In [86]:
KNNRegression(train, valid)

11971.634285714284

In [87]:
for i in range(1,100,2):
    print(f"{i} = {KNNRegression(train, valid, k=i)}")

1 = 43348.42857142857
3 = 8647.01587301587
5 = 11971.634285714284
7 = 14279.064139941689
9 = 17245.623456790134
11 = 18713.74852420307
13 = 20838.634404057473
15 = 20755.008888888897
17 = 21452.377656945137
19 = 21983.880292837355
21 = 22260.80644638808
23 = 22482.89819065623
25 = 22624.28377142857
27 = 23176.65167548501
29 = 23390.479106505856
31 = 23695.03909617957
33 = 23758.642660369933
35 = 23647.752594752183
37 = 23886.463946572057
39 = 24166.05898375129
41 = 24328.71267102915
43 = 24355.198137989646
45 = 24320.813932980593
47 = 24300.79079092026
49 = 24529.552775629196
51 = 24691.43609600703
53 = 24544.86500025427
55 = 24836.24852420307
57 = 24606.570593149543
59 = 24683.31831985882
61 = 24697.100433831143
63 = 24866.101752870465
65 = 24937.35038038884
67 = 24894.964436877453
69 = 25163.712335343713
71 = 25092.871666619434
73 = 24873.198670348233
75 = 25036.699022222223
77 = 25139.031226658328
79 = 25176.791196465765
81 = 25143.449256428685
83 = 25239.806243908522
85 = 25266.421

In [88]:
print(KNNRegression(train,test,k=3))

9727.194109772421


### Minimum mean square error happens when K = 3
### K = 3 gives 9727.19... for test set